---
title: "EJA Population Genomics Low coverage"
author: "Arthur OulÃ¨s"
format: html
editor: source
---
# Load packages

```{r}
#| label: LoadPackages
#| echo: false

c("here", "tidyverse",
  "pcadapt", "vcfR", "adegenet", "ape", "LEA", "triangulaR",
  "viridisLite") |>
  lapply(library, character.only = TRUE) |> invisible()
```

# Load data analysis functions
```{r}
#| label: LoadTidy

source(here("functions", "lcwgs_tidy_functions.R"))
source(here("functions", "plot_functions.R"))
```

# Helper objects

```{r}
#| label: GenomeMap
#| echo: false

lcwgs_vcf <- read.vcfR(here("data", "SNPS", "8-Embiotoca_filtered.vcf.gz"))

lcwgs_vcf |>
  getFIX() |>
  as_tibble() |>
  select(CHROM, POS) |>
  mutate(POS = as.numeric(POS)) |>
  saveRDS(file = here("data", "SNP_positions.rds"))

lcwgs_vcf |>
  extract_gt_tidy() |>
  _$Indiv |>
  write_rds(file = here("data", "Individual_list.rds"))

rm(lcwgs_vcf)
```

# pcadapt analysis

## All populations

### Convert vcf to bed file using plink2
```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --vcf 8-Embiotoca_filtered.vcf.gz --make-bed --out 8-Embiotoca_filtered
```

### Preanalysis to determine optimal K value
```{r}
#| label: PreAnalysis

EJA_lcwgs <- read.pcadapt(
  here("plink2", "8-Embiotoca_filtered.bed"),
  type = "bed"
)

temp <- plot(pcadapt(input = EJA_lcwgs, K = 20), option = "screeplot") + theme_classic()

ggsave(here("output", "plots", "pcadapt_all_screeplot.pdf"), temp, width = 10, height = 8)
rm(temp)
```

K = 4 seems correct

### Analysis using K = 4
```{r}
#| label: pcadapt

popmap_lcwgs <- c(
  rep("bodega_bay",            5), # BB
  rep("big_creek",            13), # BIGC -> BCR
  rep("catalina_island",      11), # CAT
  rep("elkhorn",               6), # ELK
  rep("guadalupe_island",     10), # GUA
  rep("pacific_grove",        13), # HOP -> Pacific Grove PGR 
  rep("isla_san_jeronimo",    11), # ISJ
  rep("laguna_beach",         13), # LB
  rep("point_dume",            5), # PD
  rep("santa_barbara",        13), # SB
  rep("san_clemente_island",  10), # SCL
  rep("la_jolla_san_diego",   13), # SD
  rep("tomales_bay",           1), # TB
  rep("redondo_beach",        10), # RB
  rep("santa_cruz_harbour",   10) # SCH
)

pcadapt_lcwgs <- EJA_lcwgs |> pcadapt(K = 9)
```

```{r}
#| eval: false

write_rds(pcadapt_lcwgs, file = here("output", "pcadapt_lcwgs_K9.rds"))
```

### Exploratory plots
```{r}
#| label: ExploPlots

pcadapt_lcwgs |> plot(option = "scores", pop = popmap_lcwgs, i = 1, j = 2)
pcadapt_lcwgs |> plot(option = "scores", pop = popmap_lcwgs, i = 2, j = 3)
pcadapt_lcwgs |> plot(option = "scores", pop = popmap_lcwgs, i = 3, j = 4)
pcadapt_lcwgs |> plot(option = "scores", pop = popmap_lcwgs, i = 4, j = 5)
pcadapt_lcwgs |> plot(option = "scores", pop = popmap_lcwgs, i = 5, j = 6)
pcadapt_lcwgs |> plot(option = "scores", pop = popmap_lcwgs, i = 6, j = 7)
pcadapt_lcwgs |> plot(option = "scores", pop = popmap_lcwgs, i = 7, j = 8)
pcadapt_lcwgs |> plot(option = "scores", pop = popmap_lcwgs, i = 8, j = 9)

pcadapt_lcwgs |> plot(option = "qqplot")

pcadapt_lcwgs$pvalues |> hist(
  xlab   = "p-values",
  main   = NULL,
  breaks = 50,
  col    = "orange"
)
```

## Coastal populations

### Generating whitelist

```{r}
#| label: CoastSaveWhitelist
#| eval: false

read_csv2(here("data", "individuals_coordinates.csv")) |>
  pull(Indiv) |> 
  {\(x) {
    x[!(grepl(pattern = "SCL", x = x) |
        grepl(pattern = "GUA", x = x) |
        grepl(pattern = "CAT", x = x) |
        grepl(pattern = "ISJ", x = x))]
    }}() |> 
  write_lines(here("plink2", "whitelist_coastal.txt"))
```

### Generating bed file
```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered --keep whitelist_coastal.txt --make-bed --out 8-Embiotoca_filtered_coastal
```

### Preanalysis to determine optimal K value
```{r}
#| label: CoastPreAnalysis
EJA_lcwgs_coast <- read.pcadapt(
  here("plink2", "8-Embiotoca_filtered_coastal.bed"),
  type = "bed"
)

temp <- plot(pcadapt(input = EJA_lcwgs_coast, K = 20), option = "screeplot") + theme_classic()

ggsave(here("output", "plots", "pcadapt_coast_screeplot.pdf"), temp, width = 10, height = 8)
rm(temp)
```

K = 2 seems correct

### Analysis using K = 2
```{r}
#| label: Coastpcadapt

popmap_lcwgs_coast <- c(
  rep("bodega_bay",            5), # BB
  rep("big_creek",            13), # BIGC -> BCR
  rep("elkhorn",               6), # ELK
  rep("pacific_grove",        13), # HOP -> Pacific Grove PGR
  rep("laguna_beach",         13), # LB
  rep("point_dume",            5), # PD
  rep("santa_barbara",        13), # SB
  rep("la_jolla_san_diego",   13), # SD
  rep("tomales_bay",           1), # TB
  rep("redondo_beach",        10), # RB
  rep("santa_cruz_harbour",   10)  # SCH
)

pcadapt_lcwgs_coast <- pcadapt(EJA_lcwgs_coast, K = 5)
```

```{r}
#| eval: false

pcadapt_lcwgs_coast |> 
  write_rds(file = here("output", "pcadapt_lcwgs_coast_K5.rds"))
```

### Associated exploratory plots
```{r}
#| label: CoastExploPlots

pcadapt_lcwgs_coast |> plot(option = "manhattan")

pcadapt_lcwgs_coast |>
  plot(option = "scores", pop = popmap_lcwgs_coast, i = 1, j = 2)

pcadapt_lcwgs_coast |>
  plot(option = "scores", pop = popmap_lcwgs_coast, i = 2, j = 3)

pcadapt_lcwgs_coast |>
  plot(option = "scores", pop = popmap_lcwgs_coast, i = 3, j = 4)

pcadapt_lcwgs_coast |>
  plot(option = "scores", pop = popmap_lcwgs_coast, i = 4, j = 5)

plot(pcadapt_lcwgs_coast, option = "qqplot") + theme_minimal()

pcadapt_lcwgs_coast$pvalues |> hist(
  xlab   = "p-values",
  main   = NULL,
  breaks = 50,
  col    = "orange"
)
```

```{r}
#| label: CoastMemclean

rm(EJA_lcwgs_coast, coast_PCA_1_2, coast_manhattan)
```

## North Coastal populations
### Generating bed file
```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered --keep whitelist_north.txt --make-bed --out 8-Embiotoca_filtered_north
```

### Preanalysis to determine optimal K value
```{r}
#| label: NorthPreAnalysis
EJA_lcwgs_north <- read.pcadapt(
  here("plink2", "8-Embiotoca_filtered_north.bed"),
  type = "bed"
)

pcadapt(input = EJA_lcwgs_north, K = 20) |> plot(option = "screeplot")
```

K = 3 seems correct

### Analysis using K = 3
```{r}
#| label: Northpcadapt

popmap_lcwgs_north <- c(
  rep("bodega_bay",            5), # BB
  rep("big_creek",            13), # BIGC -> BCR
  rep("elkhorn",               6), # ELK
  rep("pacific_grove",        13), # HOP -> Pacific Grove PGR
  rep("santa_barbara",        13), # SB
  rep("tomales_bay",           1), # TB
  rep("santa_cruz_harbour",   10)  # SCH
)

pcadapt_lcwgs_north <- pcadapt(EJA_lcwgs_north, K = 3)
```

```{r}
#| eval: false

write_rds(pcadapt_lcwgs_north, file = here("output", "pcadapt_lcwgs_north.rds"))
```

### Associated exploratory plots
```{r}
#| label: NorthExploPlots

north_PCA_1_2 <- pcadapt_lcwgs_north |>
  plot(option = "scores", pop = popmap_lcwgs_north, i = 1, j = 2) +
  theme_minimal()

save_open_plot(
  here("output", "plots", "north_PCA_1_2.pdf"),
  plot   = north_PCA_1_2,
  width  = 10,
  height = 8
)

plot(pcadapt_lcwgs_north, option = "qqplot") + theme_minimal()

pcadapt_lcwgs_north$pvalues |> hist(
  xlab   = "p-values",
  main   = NULL,
  breaks = 50,
  col    = "orange"
)
```

```{r}
#| label: NorthMemclean

rm(EJA_lcwgs_north, north_PCA_1_2, north_manhattan)
```

## South Coastal populations

### Generating bed file
```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered --keep whitelist_south.txt --make-bed --out 8-Embiotoca_filtered_south
```

### Preanalysis to determine optimal K value
```{r}
#| label: SouthPreAnalysis
EJA_lcwgs_south <- read.pcadapt(
  here("plink2", "8-Embiotoca_filtered_south.bed"),
  type = "bed"
)

pcadapt(input = EJA_lcwgs_south, K = 20) |> plot(option = "screeplot")
```

K = 2 seems correct

### Analysis using K = 2
```{r}
#| label: Southpcadapt

popmap_lcwgs_south <- c(
  rep("laguna_beach",         13), # LB
  rep("point_dume",            5), # PD
  rep("santa_barbara",        13), # SB
  rep("la_jolla_san_diego",   13), # SD
  rep("redondo_beach",        10) # RB
)

pcadapt_lcwgs_south <- pcadapt(EJA_lcwgs_south, K = 2)
```

```{r}
#| eval: false

write_rds(pcadapt_lcwgs_south, file = here("output", "pcadapt_lcwgs_south.rds"))
```

### Associated exploratory plots
```{r}
#| label: SouthExploPlots

south_PCA_1_2 <- pcadapt_lcwgs_south |>
  plot(option = "scores", pop = popmap_lcwgs_south, i = 1, j = 2) +
  theme_minimal()

save_open_plot(
  here("output", "plots", "south_PCA_1_2.pdf"),
  plot   = south_PCA_1_2,
  width  = 10,
  height = 8
)

plot(pcadapt_lcwgs_south, option = "qqplot") + theme_minimal()

pcadapt_lcwgs_south$pvalues |> hist(
  xlab   = "p-values",
  main   = NULL,
  breaks = 50,
  col    = "orange"
)
```

```{r}
#| label: CoastMemclean

rm(EJA_lcwgs_south, south_PCA_1_2, south_manhattan)
```

## Islands
### Generating whitelist

```{r}
#| label: ISLSaveWhitelist
#| eval: false

read_csv2(here("data", "individuals_coordinates.csv")) |>
  pull(Indiv) |> 
  {\(x) {
    x[(grepl(pattern = "CAT", x = x) |
       grepl(pattern = "GUA", x = x) |
       grepl(pattern = "ISJ", x = x) |
       grepl(pattern = "SCL", x = x))]
    }}() |> 
  write_lines(here("plink2", "whitelist_islands.txt"))
```

### Generating bed file
```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered --keep whitelist_islands.txt --make-bed --out 8-Embiotoca_filtered_isl
```

### Preanalysis to determine optimal K value
```{r}
#| label: ISLPreAnalysis
lcwgs_isl <- read.pcadapt(
  here("plink2", "8-Embiotoca_filtered_isl.bed"),
  type = "bed"
)

temp <- plot(pcadapt(input = lcwgs_isl, K = 20), option = "screeplot") + theme_classic()

ggsave(here("output", "plots", "pcadapt_islands_screeplot.pdf"), temp, width = 10, height = 8)
rm(temp)
```

K = 3 seems correct

### Analysis using K = 3
```{r}
#| label: ISLpcadapt

popmap_lcwgs_isl <- c(
  rep("catalina_island",      11), # CAT
  rep("guadalupe_island",     10), # GUA
  rep("isla_san_jeronimo",    11), # ISJ
  rep("san_clemente_island",  10)  # SCL
)

pcadapt_lcwgs_isl <- lcwgs_isl |> pcadapt(K = 3)
```

```{r}
#| eval: false

pcadapt_lcwgs_isl |>
  write_rds(file = here("output", "pcadapt_lcwgs_isl.rds"))
```

### Associated exploratory plots
```{r}
#| label: ISLExploPlots

isl_PCA_1_2 <- pcadapt_lcwgs_isl |>
  plot(option = "scores", pop = popmap_lcwgs_isl, i = 1, j = 2) +
  theme_minimal()

save_open_plot(
  here("output", "plots", "isl_PCA_1_2.pdf"),
  plot   = isl_PCA_1_2,
  width  = 10,
  height = 8
)

isl_PCA_2_3 <- pcadapt_lcwgs_isl |>
  plot(option = "scores", pop = popmap_lcwgs_isl, i = 2, j = 3) +
  theme_minimal()

save_open_plot(
  here("output", "plots", "isl_PCA_2_3.pdf"),
  plot   = isl_PCA_2_3,
  width  = 10,
  height = 8
)

pcadapt_lcwgs_isl |> plot(option = "qqplot") + theme_minimal()

pcadapt_lcwgs_isl$pvalues |> hist(
  xlab   = "p-values",
  main   = NULL,
  breaks = 50,
  col    = "orange"
)
```

```{r}
#| label: ISLMemclean

rm(lcwgs_isl, isl_PCA_1_2, isl_PCA_2_3, isl_manhattan)
```

## Coast + ISJ

```{r}
#| label: CoastISJSaveWhitelist
#| eval: false

read_csv2(here("data", "individuals_coordinates.csv")) |>
  pull(Indiv) |> 
  {\(x) {
    x[!(grepl(pattern = "SCL", x = x) |
        grepl(pattern = "GUA", x = x) |
        grepl(pattern = "CAT", x = x))]
    }}() |> 
  write_lines(here("plink2", "whitelist_coastal_isj.txt"))
```

```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered --keep whitelist_coastal_isj.txt --make-bed --out 8-Embiotoca_filtered_coastal_isj
```

```{r}
#| label: CoastISJPreAnalysis
EJA_lcwgs_coast_isj <- read.pcadapt(
  here("plink2", "8-Embiotoca_filtered_coastal_isj.bed"),
  type = "bed"
)

temp <- plot(pcadapt(input = EJA_lcwgs_coast_isj, K = 20), option = "screeplot") + theme_classic()

ggsave(here("output", "plots", "pcadapt_coast_isj_screeplot.pdf"), temp, width = 10, height = 8)
rm(temp)
```

K = 3 seems correct

```{r}
#| label: CoastISJpcadapt

popmap_lcwgs_coast_isj <- c(
  rep("bodega_bay",            5), # BB
  rep("big_creek",            13), # BIGC -> BCR
  rep("elkhorn",               6), # ELK
  rep("pacific_grove",        13), # HOP -> Pacific Grove PGR
  rep("isla_san_jeronimo",    11), # ISJ
  rep("laguna_beach",         13), # LB
  rep("point_dume",            5), # PD
  rep("santa_barbara",        13), # SB
  rep("la_jolla_san_diego",   13), # SD
  rep("tomales_bay",           1), # TB
  rep("redondo_beach",        10), # RB
  rep("santa_cruz_harbour",   10)  # SCH
)

pcadapt_lcwgs_coast_isj <- EJA_lcwgs_coast_isj |> pcadapt(K = 6)
```

```{r}
#| eval: false

pcadapt_lcwgs_coast_isj |> 
  write_rds(file = here("output", "pcadapt_lcwgs_coast_isj_K6.rds"))
```

```{r}
#| label: CoastISJExploPlots

pcadapt_lcwgs_coast_isj |>
  plot(option = "scores", pop = popmap_lcwgs_coast_isj, i = 1, j = 2)
pcadapt_lcwgs_coast_isj |>
  plot(option = "scores", pop = popmap_lcwgs_coast_isj, i = 2, j = 3)
pcadapt_lcwgs_coast_isj |>
  plot(option = "scores", pop = popmap_lcwgs_coast_isj, i = 3, j = 4)
pcadapt_lcwgs_coast_isj |>
  plot(option = "scores", pop = popmap_lcwgs_coast_isj, i = 4, j = 5)
pcadapt_lcwgs_coast_isj |>
  plot(option = "scores", pop = popmap_lcwgs_coast_isj, i = 5, j = 6)

pcadapt_lcwgs_coast_isj |> plot(option = "qqplot") + theme_minimal()

pcadapt_lcwgs_coast_isj$pvalues |> hist(
  xlab   = "p-values",
  main   = NULL,
  breaks = 50,
  col    = "orange"
)
```

```{r}
#| label: CoastMemclean

rm(EJA_lcwgs_coast_isj, coast_isj_PCA_1_2, coast_isj_PCA_2_3, coast_isj_manhattan)
```

## Coast + SCL

```{r}
#| label: CoastSCLSaveWhitelist
#| eval: false

read_csv2(here("data", "individuals_coordinates.csv")) |>
  pull(Indiv) |> 
  {\(x) {
    x[!(grepl(pattern = "ISJ", x = x) |
        grepl(pattern = "GUA", x = x) |
        grepl(pattern = "CAT", x = x))]
    }}() |> 
  write_lines(here("plink2", "whitelist_coastal_scl.txt"))
```

```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered --keep whitelist_coastal_scl.txt --make-bed --out 8-Embiotoca_filtered_coastal_scl
```

```{r}
#| label: CoastSCLPreAnalysis
lcwgs_coast_scl <- read.pcadapt(
  here("plink2", "8-Embiotoca_filtered_coastal_scl.bed"),
  type = "bed"
)

temp <- plot(pcadapt(input = lcwgs_coast_scl, K = 20), option = "screeplot") + theme_classic()

ggsave(here("output", "plots", "pcadapt_coast_scl_screeplot.pdf"), temp, width = 10, height = 8)
rm(temp)
```

K = 3 seems correct

```{r}
#| label: CoastSCLpcadapt

popmap_lcwgs_coast_scl <- c(
  rep("bodega_bay",            5), # BB
  rep("big_creek",            13), # BIGC -> BCR
  rep("elkhorn",               6), # ELK
  rep("pacific_grove",        13), # HOP -> Pacific Grove PGR
  rep("laguna_beach",         13), # LB
  rep("point_dume",            5), # PD
  rep("santa_barbara",        13), # SB
  rep("san_clemente_island",  10), # SCL
  rep("la_jolla_san_diego",   13), # SD
  rep("tomales_bay",           1), # TB
  rep("redondo_beach",        10), # RB
  rep("santa_cruz_harbour",   10)  # SCH
)

pcadapt_lcwgs_coast_scl <- lcwgs_coast_scl |> pcadapt(K = 6)
```

```{r}
#| eval: false

pcadapt_lcwgs_coast_scl |> 
  write_rds(file = here("output", "pcadapt_lcwgs_coast_scl_K6.rds"))
```

```{r}
#| label: CoastSCLExploPlots

pcadapt_lcwgs_coast_scl |>
  plot(option = "scores", pop = popmap_lcwgs_coast_scl, i = 1, j = 2)

pcadapt_lcwgs_coast_scl |>
  plot(option = "scores", pop = popmap_lcwgs_coast_scl, i = 2, j = 3)

pcadapt_lcwgs_coast_scl |>
  plot(option = "scores", pop = popmap_lcwgs_coast_scl, i = 3, j = 4)

pcadapt_lcwgs_coast_scl |>
  plot(option = "scores", pop = popmap_lcwgs_coast_scl, i = 4, j = 5)

pcadapt_lcwgs_coast_scl |>
  plot(option = "scores", pop = popmap_lcwgs_coast_scl, i = 5, j = 6)

pcadapt_lcwgs_coast_scl |> plot(option = "qqplot") + theme_minimal()

pcadapt_lcwgs_coast_scl$pvalues |> hist(
  xlab   = "p-values",
  main   = NULL,
  breaks = 50,
  col    = "orange"
)
```

```{r}
#| label: CoastSCLMemclean

rm(lcwgs_coast_scl, coast_scl_PCA_1_2, coast_scl_PCA_2_3, coast_scl_manhattan)
```

## Coast + GUA

```{r}
#| label: CoastGUASaveWhitelist
#| eval: false

read_csv2(here("data", "individuals_coordinates.csv")) |>
  pull(Indiv) |> 
  {\(x) {
    x[!(grepl(pattern = "ISJ", x = x) |
        grepl(pattern = "SCL", x = x) |
        grepl(pattern = "CAT", x = x))]
    }}() |> 
  write_lines(here("plink2", "whitelist_coastal_gua.txt"))
```

```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered --keep whitelist_coastal_gua.txt --make-bed --out 8-Embiotoca_filtered_coastal_gua
```

```{r}
#| label: CoastGUAPreAnalysis
lcwgs_coast_gua <- read.pcadapt(
  here("plink2", "8-Embiotoca_filtered_coastal_gua.bed"),
  type = "bed"
)

temp <- plot(pcadapt(input = lcwgs_coast_gua, K = 20), option = "screeplot") + theme_classic()

ggsave(here("output", "plots", "pcadapt_coast_gua_screeplot.pdf"), temp, width = 10, height = 8)
rm(temp)
```

K = 3 seems correct

```{r}
#| label: CoastGUApcadapt

popmap_lcwgs_coast_gua <- c(
  rep("bodega_bay",            5), # BB
  rep("big_creek",            13), # BIGC -> BCR
  rep("elkhorn",               6), # ELK
  rep("guadalupe_island",     10), # GUA
  rep("pacific_grove",        13), # HOP -> Pacific Grove PGR
  rep("laguna_beach",         13), # LB
  rep("point_dume",            5), # PD
  rep("santa_barbara",        13), # SB
  rep("la_jolla_san_diego",   13), # SD
  rep("tomales_bay",           1), # TB
  rep("redondo_beach",        10), # RB
  rep("santa_cruz_harbour",   10)  # SCH
)

pcadapt_lcwgs_coast_gua <- lcwgs_coast_gua |> pcadapt(K = 6)
```

```{r}
#| eval: false

pcadapt_lcwgs_coast_gua |>
  write_rds(file = here("output", "pcadapt_lcwgs_coast_gua_K6.rds"))
```

```{r}
#| label: CoastGUAExploPlots

coast_gua_PCA_1_2 <- pcadapt_lcwgs_coast_gua |>
  plot(option = "scores", pop = popmap_lcwgs_coast_gua, i = 1, j = 2)

coast_gua_PCA_2_3 <- pcadapt_lcwgs_coast_gua |>
  plot(option = "scores", pop = popmap_lcwgs_coast_gua, i = 2, j = 3)

coast_gua_PCA_2_3 <- pcadapt_lcwgs_coast_gua |>
  plot(option = "scores", pop = popmap_lcwgs_coast_gua, i = 3, j = 4)

coast_gua_PCA_2_3 <- pcadapt_lcwgs_coast_gua |>
  plot(option = "scores", pop = popmap_lcwgs_coast_gua, i = 4, j = 5)

coast_gua_PCA_2_3 <- pcadapt_lcwgs_coast_gua |>
  plot(option = "scores", pop = popmap_lcwgs_coast_gua, i = 5, j = 6)

pcadapt_lcwgs_coast_gua |> plot(option = "qqplot") + theme_minimal()

pcadapt_lcwgs_coast_gua$pvalues |> hist(
  xlab   = "p-values",
  main   = NULL,
  breaks = 50,
  col    = "orange"
)
```

```{r}
#| label: CoastGUAMemclean

rm(lcwgs_coast_gua, coast_gua_PCA_1_2, coast_gua_PCA_2_3, coast_gua_manhattan)
```

## Coast + CAT

```{r}
#| label: CoastCATSaveWhitelist
#| eval: false

read_csv2(here("data", "individuals_coordinates.csv")) |>
  pull(Indiv) |> 
  {\(x) {
    x[!(grepl(pattern = "ISJ", x = x) |
        grepl(pattern = "SCL", x = x) |
        grepl(pattern = "GUA", x = x))]
    }}() |> 
  write_lines(here("plink2", "whitelist_coastal_cat.txt"))
```

```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered --keep whitelist_coastal_cat.txt --make-bed --out 8-Embiotoca_filtered_coastal_cat
```

```{r}
#| label: CoastCATPreAnalysis
lcwgs_coast_cat <- read.pcadapt(
  here("plink2", "8-Embiotoca_filtered_coastal_cat.bed"),
  type = "bed"
)

temp <- plot(pcadapt(input = lcwgs_coast_cat, K = 20), option = "screeplot") + theme_classic()

ggsave(here("output", "plots", "pcadapt_coast_cat_screeplot.pdf"), temp, width = 10, height = 8)
rm(temp)
```

K = 3 seems correct

```{r}
#| label: CoastCATpcadapt

popmap_lcwgs_coast_cat <- c(
  rep("bodega_bay",            5), # BB
  rep("big_creek",            13), # BIGC -> BCR
  rep("catalina_island",      11), # CAT
  rep("elkhorn",               6), # ELK
  rep("pacific_grove",        13), # HOP -> Pacific Grove PGR
  rep("laguna_beach",         13), # LB
  rep("point_dume",            5), # PD
  rep("santa_barbara",        13), # SB
  rep("la_jolla_san_diego",   13), # SD
  rep("tomales_bay",           1), # TB
  rep("redondo_beach",        10), # RB
  rep("santa_cruz_harbour",   10)  # SCH
)

pcadapt_lcwgs_coast_cat <- lcwgs_coast_cat |> pcadapt(K = 6)
```

```{r}
#| eval: false

pcadapt_lcwgs_coast_cat |> 
  write_rds(file = here("output", "pcadapt_lcwgs_coast_cat_K6.rds"))
```

```{r}
#| label: CoastCATExploPlots

pcadapt_lcwgs_coast_cat |>
  plot(option = "scores", pop = popmap_lcwgs_coast_cat, i = 1, j = 2)

pcadapt_lcwgs_coast_cat |>
  plot(option = "scores", pop = popmap_lcwgs_coast_cat, i = 2, j = 3)
  
pcadapt_lcwgs_coast_cat |>
  plot(option = "scores", pop = popmap_lcwgs_coast_cat, i = 3, j = 4)

pcadapt_lcwgs_coast_cat |>
  plot(option = "scores", pop = popmap_lcwgs_coast_cat, i = 4, j = 5)

pcadapt_lcwgs_coast_cat |>
  plot(option = "scores", pop = popmap_lcwgs_coast_cat, i = 5, j = 6)


pcadapt_lcwgs_coast_cat |> plot(option = "qqplot") + theme_minimal()

pcadapt_lcwgs_coast_cat$pvalues |> hist(
  xlab   = "p-values",
  main   = NULL,
  breaks = 50,
  col    = "orange"
)
```

```{r}
#| label: CoastCATMemclean

rm(lcwgs_coast_cat, coast_cat_PCA_1_2, coast_cat_PCA_2_3, coast_cat_manhattan)
```

## CAT + SCL

```{r}
#| label: CATSCLSaveWhitelist
#| eval: false

read_csv2(here("data", "individuals_coordinates.csv")) |>
  pull(Indiv) |> 
  {\(x) {
    x[(grepl(pattern = "SCL", x = x) |
       grepl(pattern = "CAT", x = x))]
    }}() |> 
  write_lines(here("plink2", "whitelist_cat_scl.txt"))
```

```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered --keep whitelist_cat_scl.txt --make-bed --out 8-Embiotoca_filtered_cat_scl
```

```{r}
#| label: CATSCLPreAnalysis

lcwgs_cat_scl <- read.pcadapt(
  here("plink2", "8-Embiotoca_filtered_cat_scl.bed"),
  type = "bed"
)

temp <- plot(pcadapt(input = lcwgs_cat_scl, K = 20), option = "screeplot") + theme_classic()

ggsave(here("output", "plots", "pcadapt_cat_scl_screeplot.pdf"), temp, width = 10, height = 8)
rm(temp)
```

K = 1 seems correct

```{r}
#| label: CATSCLpcadapt

popmap_lcwgs_cat_scl <- c(
  rep("catalina_island",      11), # CAT
  rep("san_clemente_island",  10)  # SCL
)

pcadapt_lcwgs_cat_scl <- lcwgs_cat_scl |> pcadapt(K = 1)
```

```{r}
#| eval: false

pcadapt_lcwgs_cat_scl |> 
  write_rds(file = here("output", "pcadapt_lcwgs_cat_scl.rds"))
```

```{r}
#| label: CATSCLExploPlots

pcadapt_lcwgs_cat_scl |> plot(option = "qqplot") + theme_minimal()

pcadapt_lcwgs_cat_scl$pvalues |> hist(
  xlab   = "p-values",
  main   = NULL,
  breaks = 50,
  col    = "orange"
)
```

```{r}
#| label: CoastCATMemclean

rm(lcwgs_cat_scl, cat_scl_manhattan)
```

## GUA + CAT

```{r}
#| label: GUACATSaveWhitelist
#| eval: false

read_csv2(here("data", "individuals_coordinates.csv")) |>
  pull(Indiv) |> 
  {\(x) {
    x[(grepl(pattern = "GUA", x = x) |
       grepl(pattern = "CAT", x = x))]
    }}() |> 
  write_lines(here("plink2", "whitelist_gua_cat.txt"))
```

```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered --keep whitelist_gua_cat.txt --make-bed --out 8-Embiotoca_filtered_gua_cat
```

```{r}
#| label: GUACATPreAnalysis

lcwgs_gua_cat <- read.pcadapt(
  here("plink2", "8-Embiotoca_filtered_gua_cat.bed"),
  type = "bed"
)

temp <- plot(pcadapt(input = lcwgs_gua_cat, K = 20), option = "screeplot") + theme_classic()

ggsave(here("output", "plots", "pcadapt_gua_cat_screeplot.pdf"), temp, width = 10, height = 8)
rm(temp)
```

K = 1 seems correct

```{r}
#| label: GUACATpcadapt

popmap_lcwgs_gua_cat <- c(
  rep("catalina_island",      11), # CAT
  rep("guadalupe_island",     10)  # GUA
)

pcadapt_lcwgs_gua_cat <- lcwgs_gua_cat |> pcadapt(K = 1)
```

```{r}
#| eval: false

pcadapt_lcwgs_gua_cat |> 
  write_rds(file = here("output", "pcadapt_lcwgs_gua_cat.rds"))
```

```{r}
#| label: GUACATExploPlots

pcadapt_lcwgs_gua_cat |> plot(option = "qqplot") + theme_minimal()

pcadapt_lcwgs_gua_cat$pvalues |> hist(
  xlab   = "p-values",
  main   = NULL,
  breaks = 50,
  col    = "orange"
)
```

```{r}
#| label: GUACATMemclean

rm(lcwgs_gua_cat, gua_cat_manhattan)
```

## GUA + SCL

```{r}
#| label: GUASCLSaveWhitelist
#| eval: false

read_csv2(here("data", "individuals_coordinates.csv")) |>
  pull(Indiv) |> 
  {\(x) {
    x[(grepl(pattern = "SCL", x = x) |
       grepl(pattern = "GUA", x = x))]
    }}() |> 
  write_lines(here("plink2", "whitelist_gua_scl.txt"))
```

```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered --keep whitelist_gua_scl.txt --make-bed --out 8-Embiotoca_filtered_gua_scl
```

```{r}
#| label: GUASCLPreAnalysis

lcwgs_gua_scl <- read.pcadapt(
  here("plink2", "8-Embiotoca_filtered_gua_scl.bed"),
  type = "bed"
)

temp <- plot(pcadapt(input = lcwgs_gua_scl, K = 19), option = "screeplot") + theme_classic()

ggsave(here("output", "plots", "pcadapt_gua_scl_screeplot.pdf"), temp, width = 10, height = 8)
rm(temp)
```

K = 1 seems correct

```{r}
#| label: GUASCLpcadapt

popmap_lcwgs_gua_scl <- c(
  rep("guadalupe_island",     10), # GUA
  rep("san_clemente_island",  10)  # SCL
)

pcadapt_lcwgs_gua_scl <- lcwgs_gua_scl |> pcadapt(K = 1)
```

```{r}
#| eval: false

pcadapt_lcwgs_gua_scl |> 
  write_rds(file = here("output", "pcadapt_lcwgs_gua_scl.rds"))
```

```{r}
#| label: GUASCLExploPlots

pcadapt_lcwgs_gua_scl |> plot(option = "qqplot") + theme_minimal()

pcadapt_lcwgs_gua_scl$pvalues |> hist(
  xlab   = "p-values",
  main   = NULL,
  breaks = 50,
  col    = "orange"
)
```

```{r}
#| label: GUASCLMemclean

rm(lcwgs_gua_scl, gua_scl_manhattan)
```

## GUA + ISJ

```{r}
#| label: GUAISJSaveWhitelist
#| eval: false

read_csv2(here("data", "individuals_coordinates.csv")) |>
  pull(Indiv) |> 
  {\(x) {
    x[(grepl(pattern = "GUA", x = x) |
       grepl(pattern = "ISJ", x = x))]
    }}() |> 
  write_lines(here("plink2", "whitelist_gua_isj.txt"))
```

```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered --keep whitelist_GUA_ISJ.txt --make-bed --out 8-Embiotoca_filtered_GUA_ISJ
```

```{r}
#| label: GUAISJPreAnalysis

lcwgs_gua_isj <- read.pcadapt(
  here("plink2", "8-Embiotoca_filtered_gua_isj.bed"),
  type = "bed"
)

temp <- plot(pcadapt(input = lcwgs_gua_isj, K = 19), option = "screeplot") + theme_classic()

ggsave(here("output", "plots", "pcadapt_gua_isj_screeplot.pdf"), temp, width = 10, height = 8)
rm(temp)
```

K = 1 seems correct

```{r}
#| label: GUAISJpcadapt

popmap_lcwgs_GUA_ISJ <- c(
  rep("guadalupe_island",     10), # GUA
  rep("isla_san_jeronimo",    11)  # ISJ
)

pcadapt_lcwgs_GUA_ISJ <- lcwgs_gua_isj |> pcadapt(K = 1)
```

```{r}
#| eval: false

pcadapt_lcwgs_GUA_ISJ |> 
  write_rds(file = here("output", "pcadapt_lcwgs_gua_isj.rds"))
```

```{r}
#| label: GUAISJExploPlots

pcadapt_lcwgs_GUA_ISJ |> plot(option = "qqplot") + theme_minimal()

pcadapt_lcwgs_GUA_ISJ$pvalues |> hist(
  xlab   = "p-values",
  main   = NULL,
  breaks = 50,
  col    = "orange"
)
```

```{r}
#| label: GUAISJMemclean

rm(lcwgs_GUA_ISJ, GUA_ISJ_manhattan)
```

## CAT + ISJ

```{r}
#| label: CATISJSaveWhitelist
#| eval: false

read_csv2(here("data", "individuals_coordinates.csv")) |>
  pull(Indiv) |> 
  {\(x) {
    x[(grepl(pattern = "CAT", x = x) |
       grepl(pattern = "ISJ", x = x))]
    }}() |> 
  write_lines(here("plink2", "whitelist_cat_isj.txt"))
```

```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered --keep whitelist_cat_isj.txt --make-bed --out 8-Embiotoca_filtered_cat_isj
```

```{r}
#| label: CATISJPreAnalysis

lcwgs_cat_isj <- read.pcadapt(
  here("plink2", "8-Embiotoca_filtered_cat_isj.bed"),
  type = "bed"
)

temp <- plot(pcadapt(input = lcwgs_cat_isj, K = 19), option = "screeplot") + theme_classic()

ggsave(here("output", "plots", "pcadapt_cat_isj_screeplot.pdf"), temp, width = 10, height = 8)
rm(temp)
```

K = 1 seems correct

```{r}
#| label: CATISJpcadapt

popmap_lcwgs_cat_isj <-  c(
  rep("catalina_island",      11), # CAT
  rep("isla_san_jeronimo",    11)  # ISJ
)

pcadapt_lcwgs_cat_isj <- lcwgs_cat_isj |> pcadapt(K = 1)
```

```{r}
#| eval: false

pcadapt_lcwgs_cat_isj |> 
  write_rds(file = here("output", "pcadapt_lcwgs_cat_isj.rds"))
```

```{r}
#| label: CATISJExploPlots

pcadapt_lcwgs_cat_isj |> plot(option = "qqplot") + theme_minimal()

pcadapt_lcwgs_cat_isj$pvalues |> hist(
  xlab   = "p-values",
  main   = NULL,
  breaks = 50,
  col    = "orange"
)
```

```{r}
#| label: CATISJMemclean

rm(lcwgs_cat_isj, cat_isj_manhattan)
```

## SCL + ISJ

```{r}
#| label: SCLISJSaveWhitelist
#| eval: false

read_csv2(here("data", "individuals_coordinates.csv")) |>
  pull(Indiv) |> 
  {\(x) {
    x[(grepl(pattern = "ISJ", x = x) |
       grepl(pattern = "SCL", x = x))]
    }}() |> 
  write_lines(here("plink2", "whitelist_scl_isj.txt"))
```

```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered --keep whitelist_scl_isj.txt --make-bed --out 8-Embiotoca_filtered_scl_isj
```

```{r}
#| label: SCLISJPreAnalysis

lcwgs_scl_isj <- read.pcadapt(
  here("plink2", "8-Embiotoca_filtered_scl_isj.bed"),
  type = "bed"
)

temp <- plot(pcadapt(input = lcwgs_scl_isj, K = 19), option = "screeplot") + theme_classic()

ggsave(here("output", "plots", "pcadapt_scl_isj_screeplot.pdf"), temp, width = 10, height = 8)
rm(temp)
```

K = 1 seems correct

```{r}
#| label: SCLISJpcadapt

popmap_lcwgs_scl_isj <- c(
  rep("isla_san_jeronimo",    11), # ISJ
  rep("san_clemente_island",  10)  # SCL
)

pcadapt_lcwgs_scl_isj <- lcwgs_scl_isj |> pcadapt(K = 1)
```

```{r}
#| eval: false

pcadapt_lcwgs_scl_isj |> 
  write_rds(file = here("output", "pcadapt_lcwgs_scl_isj.rds"))
```

```{r}
#| label: SCLISJExploPlots

pcadapt_lcwgs_scl_isj |> plot(option = "qqplot") + theme_minimal()

pcadapt_lcwgs_scl_isj$pvalues |> hist(
  xlab   = "p-values",
  main   = NULL,
  breaks = 50,
  col    = "orange"
)
```

```{r}
#| label: SCLISJMemclean

rm(lcwgs_scl_isj, scl_isj_manhattan)
```

# Outliers

## Load species Metadata

```{r}
EJA_annotations <- here("data", "EJA_annotations.gff") |> # Load annotated genome
  read.gff() |>
  as_tibble()

chromosomes_info <- here("data", "JAKOON01_catalog", "sequence_report.tsv") |>  # Load misc inform
  read_tsv() |>
  select(c(`GenBank seq accession`, `RefSeq seq accession`))
```

### All outliers

```{r}
read_write_locations <- tibble(
  read_files = list.files(here("output")) |> str_subset("pcadapt")
) |> 
  mutate(
    write_files = read_files |>
      str_replace("pcadapt", "EJA") |>
      str_replace(".rds", "_5E-2_outliers_annotations.csv")
  )

map2(
  read_write_locations$read_files, read_write_locations$write_files,
     \(x, y) {
       print(x)
       
       here("output", x) |>
         readRDS() |>
         Get_outliers(method = "bonferroni", pvalue_threshold = .05) |> 
         Get_annotations(chrom_info = chromosomes_info, gff_annotation = EJA_annotations) |>
         Clean_annotations() |> 
         write_csv2(here("output", y))
     }
)
```
####Log:

[1] "pcadapt_lcwgs_cat_isj.rds"
[1] "There are 48182 outliers SNPs."
[1] "Outliers number was reduced to 10.000 lowest pvalues."
[1] "There are 10000 annotated outliers SNPs."
[1] "There are 2886 genes."
[1] "There are 6882 rnas."
[1] "There are 1704 exons."
[1] "There are 754 cds."
[1] "There are 3145 unique SNPs."

[1] "pcadapt_lcwgs_cat_scl.rds"
[1] "There are 8467 outliers SNPs."
[1] "There are 8467 annotated outliers SNPs."
[1] "There are 1879 genes."
[1] "There are 4252 rnas."
[1] "There are 1040 exons."
[1] "There are 457 cds."
[1] "There are 2063 unique SNPs."

[1] "pcadapt_lcwgs_coast_cat_K6.rds"
[1] "There are 13850 outliers SNPs."
[1] "Outliers number was reduced to 10.000 lowest pvalues."
[1] "There are 10000 annotated outliers SNPs."
[1] "There are 2764 genes."
[1] "There are 6261 rnas."
[1] "There are 1510 exons."
[1] "There are 656 cds."
[1] "There are 3005 unique SNPs."

[1] "pcadapt_lcwgs_coast_gua_K6.rds"
[1] "There are 23112 outliers SNPs."
[1] "Outliers number was reduced to 10.000 lowest pvalues."
[1] "There are 10000 annotated outliers SNPs."
[1] "There are 2631 genes."
[1] "There are 6095 rnas."
[1] "There are 1976 exons."
[1] "There are 752 cds."
[1] "There are 2890 unique SNPs."

[1] "pcadapt_lcwgs_coast_isj_K6.rds"
[1] "There are 12536 outliers SNPs."
[1] "Outliers number was reduced to 10.000 lowest pvalues."
[1] "There are 10000 annotated outliers SNPs."
[1] "There are 1786 genes."
[1] "There are 3988 rnas."
[1] "There are 1517 exons."
[1] "There are 740 cds."
[1] "There are 2098 unique SNPs."

[1] "pcadapt_lcwgs_coast_K5.rds"
[1] "There are 4507 outliers SNPs."
[1] "There are 4507 annotated outliers SNPs."
[1] "There are 1261 genes."
[1] "There are 2794 rnas."
[1] "There are 695 exons."
[1] "There are 339 cds."
[1] "There are 1350 unique SNPs."

[1] "pcadapt_lcwgs_coast_scl_K6.rds"
[1] "There are 17167 outliers SNPs."
[1] "Outliers number was reduced to 10.000 lowest pvalues."
[1] "There are 10000 annotated outliers SNPs."
[1] "There are 2968 genes."
[1] "There are 6757 rnas."
[1] "There are 1769 exons."
[1] "There are 843 cds."
[1] "There are 3233 unique SNPs."

[1] "pcadapt_lcwgs_gua_cat.rds"
[1] "There are 38986 outliers SNPs."
[1] "Outliers number was reduced to 10.000 lowest pvalues."
[1] "There are 10000 annotated outliers SNPs."
[1] "There are 1940 genes."
[1] "There are 4274 rnas."
[1] "There are 1145 exons."
[1] "There are 493 cds."
[1] "There are 2210 unique SNPs."

[1] "pcadapt_lcwgs_gua_isj.rds"
[1] "There are 75834 outliers SNPs."
[1] "Outliers number was reduced to 10.000 lowest pvalues."
[1] "There are 10000 annotated outliers SNPs."
[1] "There are 2533 genes."
[1] "There are 5744 rnas."
[1] "There are 1551 exons."
[1] "There are 648 cds."
[1] "There are 2807 unique SNPs."

[1] "pcadapt_lcwgs_gua_scl.rds"
[1] "There are 44863 outliers SNPs."
[1] "Outliers number was reduced to 10.000 lowest pvalues."
[1] "There are 10000 annotated outliers SNPs."
[1] "There are 1850 genes."
[1] "There are 4214 rnas."
[1] "There are 1259 exons."
[1] "There are 535 cds."
[1] "There are 2154 unique SNPs."

[1] "pcadapt_lcwgs_isl.rds"
[1] "There are 55838 outliers SNPs."
[1] "Outliers number was reduced to 10.000 lowest pvalues."
[1] "There are 10000 annotated outliers SNPs."
[1] "There are 2854 genes."
[1] "There are 6643 rnas."
[1] "There are 1653 exons."
[1] "There are 770 cds."
[1] "There are 3119 unique SNPs."

[1] "pcadapt_lcwgs_K9.rds"
[1] "There are 56775 outliers SNPs."
[1] "Outliers number was reduced to 10.000 lowest pvalues."
[1] "There are 10000 annotated outliers SNPs."
[1] "There are 2915 genes."
[1] "There are 6733 rnas."
[1] "There are 1997 exons."
[1] "There are 837 cds."
[1] "There are 3219 unique SNPs."

[1] "pcadapt_lcwgs_scl_isj.rds"
[1] "There are 64969 outliers SNPs."
[1] "Outliers number was reduced to 10.000 lowest pvalues."
[1] "There are 10000 annotated outliers SNPs."
[1] "There are 2692 genes."
[1] "There are 6394 rnas."
[1] "There are 1745 exons."
[1] "There are 770 cds."
[1] "There are 2977 unique SNPs."

## For only one

```{r}
here("output", "pcadapt_lcwgs_gua_isj.rds") |>
  readRDS() |>
  Get_outliers(method = "bonferroni", pvalue_threshold = .05) |> 
  Get_annotations(chrom_info = chromosomes_info, gff_annotation = EJA_annotations) |>
  Clean_annotations() |> 
  write_csv2(here("output", "EJA_lcwgs_gua_isj_5E-2_outliers_annotations.csv"))
```

# Allele frequencies

## Send-away code

```{r}
#| eval: false

EJA_gen <- readRDS(here("data", "EJA_lcwgs_gen.rds"))

EJA_allele_frequencies <- makefreq(EJA_gen)

EJA_allele_frequencies_tb <- EJA_allele_frequencies |>
    t() |>
    as_tibble(rownames = "POSITION") |>
    separate_wider_delim(POSITION, delim = "_", names = c("CHROM", "TBD", "POSALLELE")) |>
    separate_wider_delim(POSALLELE, delim = ".", names = c("POS", "ALLELE")) |>
    select(-c(TBD)) |>
    mutate(CHROM = paste0(CHROM, ".1"))

saveRDS(EJA_allele_frequencies_tb, here("output", "EJA_allele_frequencies_tb.rds"))
```

## Results exploration

```{r}
allele_freq <- readRDS(here("output", "EJA_allele_frequencies_tb.rds"))
allele_freq |> write_csv2(here("output", "EJA_allele_frequencies_tb.csv"))
```

```{r}
allele_freq |>
  filter(CHROM == "JAKOON010000124.1")
```


# Admixture analysis

## Convert data to ped format for LEA import using plink2
```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2
./plink2 --vcf 8-Embiotoca_filtered.vcf.gz --export ped --out 8-Embiotoca_filtered
```

## Read and convert data
```{r}
#| label: VcfConvert
#| eval: false

ped2geno(
  input.file  = here("plink2", "8-Embiotoca_filtered.ped"),
  output.file = here("data", "SNPs", "8-Embiotoca_filtered.geno")
)
```

## Run analysis
```{r}
#| label: snmfAnalysisSingle

snmf_analysis <- snmf(
  here("data", "SNPs", "8-Embiotoca_filtered.geno"),
  K           = 4:8,
  project     = "continue",
  repetitions = 50,
  entropy     = TRUE
)
```

## Plot cross entropy to chose optimal value of K
```{r}
#| label: snmfCrossEntropy

plot(snmf_analysis)

snmf_tb <- tibble(K = as.numeric(unique(snmf_analysis@K)))

cross_entropy <- rep(0, dim(snmf_tb)[1])

for (i in 1:length(snmf_tb$K)) {
  cross_entropy[i] <- min(cross.entropy(snmf_analysis, K = snmf_tb$K[i]))
}

snmf_plot <- snmf_tb |>
  mutate(cross_entropy = cross_entropy) |> 
  ggplot() +
  geom_point(aes(x = K, y = cross_entropy)) +
  theme_classic()

ggsave(
  filename = here("output", "plots", "snmf_cross_entropy.png"),
  plot     = snmf_plot,
  width    = 8,
  height   = 5
)
```

K = 7 has the minimum cross-entropy.

## Plot results for K = 7
```{r}
#| label: snmfBarPlot

barchart(
  snmf_analysis,
  K         = 7,
  run       = which.min(cross.entropy(snmf_analysis, K = 7)),
  col       = viridis::turbo(n = 6),
  sort.by.Q = FALSE
)
```

# Triangle plots

## For North-South Gradient

```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered_coastal --keep-allele-order --recode vcf --out 8-Embiotoca_filtered_coastal
```

```{r}
lcwgs_coast_vcf <- read.vcfR(here("plink2", "8-Embiotoca_filtered_coastal.vcf"))

popmap_lcwgs_coast <- read_csv2(here("data", "individuals_coordinates.csv")) |> 
  select(Indiv, Localisation) |>
  rename(id = Indiv, pop = Localisation) |> 
  filter(pop != c("CAT")) |> 
  filter(pop != c("ISJ")) |>
  filter(pop != c("SCL")) |> 
  filter(pop != c("GUA")) |> 
  data.frame()
```

```{r}
lcwgs_coast_vcf_diffs <- vector(mode = 'list', length = 3)
parameters <- c(.5, .75, 1)

for (i in 1:3) {
  lcwgs_coast_vcf_diffs[[i]] <- alleleFreqDiff(lcwgs_coast_vcf,
                                               pm = popmap_lcwgs_coast,
                                               p1 = "BBA", p2 = "SD",
                                               difference = parameters[i])
}

rm(lcwgs_coast_vcf)
```
with 1 difference : "8456 sites passed allele frequency difference threshold"

Difference = .75 76178

Difference = .5 : 276405 sites

```{r}
hi_het_coast <- vector(mode = 'list', length = 3)

for (i in 1:3) {
  hi_het_coast[[i]] <- hybridIndex(lcwgs_coast_vcf_diffs[[i]],
                                   pm = popmap_lcwgs_coast,
                                   p1 = "BBA", p2 = "SD")
}
```

```{r}
#| eval: false

write_rds(hi_het_coast, here("output", "Triangle_analysis_coast.rds"))
```

```{r}
hi_het_coast_plots <- vector(mode = 'list', length = 3)

for (i in 1:3) {
  hi_het_coast_plots[[i]] <- hi_het_coast[[i]] |>
    mutate(
      pop = factor(
        pop,
        levels = c("BBA", "TB", "SCH", "ELK", "HOP",
                   "BIGC", "SB", "PD", "RB", "LAG", "SD")
      )
    ) |> 
    arrange(pop) |>
    triangle.plot(colors = turbo(n = 11)) +
    labs(title = paste0("d = ", as.character(parameters[i])))
}

save_open_plot(
  path   = here("output", "plots", "coast_Triangle_plots.pdf"),
  plot   = Reduce("+", hi_het_coast_plots) +
    plot_layout(axis_titles = "collect", guides = "collect"),
  width  = 14,
  height = 4
)
```

```{r}
rm(lcwgs_coast_vcf_diffs, hi_het_coast, hi_het_coast_plots)
gc()
```


## For Coastal + ISJ

```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered_coastal_isj --keep-allele-order --recode vcf --out 8-Embiotoca_filtered_coastal_isj
```

```{r}
EJA_lcwgs_coast_isj_vcf <- read.vcfR(here("plink2", "8-Embiotoca_filtered_coastal_isj.vcf"))

popmap_lcwgs_coast_isj <- read_csv2(here("data", "individuals_coordinates.csv")) |> 
  select(Indiv, Localisation) |>
  rename(id = Indiv, pop = Localisation) |> 
  filter(pop != c("CAT")) |> 
  filter(pop != c("SCL")) |> 
  filter(pop != c("GUA")) |> 
  data.frame()
```

```{r}
EJA_lcwgs_coast_isj_vcf_diff <- vector(mode='list', length=3)
parameters <- c(.5, .75, 1)

for (i in 1:3) {
  EJA_lcwgs_coast_isj_vcf_diff[[i]] <- alleleFreqDiff(EJA_lcwgs_coast_isj_vcf,
                                                      pm = popmap_lcwgs_coast_isj,
                                                      p1 = "BBA", p2 = "ISJ",
                                                      difference = parameters[i])
}

rm(EJA_lcwgs_coast_isj_vcf)
gc()
```
With 1 : "32484 sites passed allele frequency difference threshold"
With .75 : "166077 sites passed allele frequency difference threshold"
With .5 : "424389 sites passed allele frequency difference threshold"

```{r}
hi_het_coast_isj <- vector(mode='list', length=3)

for (i in 1:3) {
  hi_het_coast_isj[[i]] <- hybridIndex(EJA_lcwgs_coast_isj_vcf_diff[[i]],
                                       pm = popmap_lcwgs_coast_isj,
                                       p1 = "BBA", p2 = "ISJ")
}
```

```{r}
#| eval: false

write_rds(hi_het_coast_isj, here("output", "Triangle_analysis_coast_isj.rds"))
```

```{r}
hi_het_coast_isj_plots <- vector(mode='list', length=3)

for (i in 1:3) {
  hi_het_coast_isj_plots[[i]] <- hi_het_coast_isj[[i]] |>
    mutate(
      pop = factor(
        pop,
        levels = c("BBA", "TB", "SCH", "ELK", "HOP", "BIGC",
                   "SB",  "PD", "RB",  "LAG", "SD",  "ISJ")
      )
    ) |> 
    arrange(pop) |>
    triangle.plot(colors = turbo(n = 12)) +
    labs(title = paste0("d = ", as.character(parameters[i])))
}

save_open_plot(
  path   = here("output", "plots", "coast_isj_Triangle_plots.pdf"),
  plot   = Reduce("+", hi_het_coast_isj_plots) +
    plot_layout(axis_titles = "collect", guides = "collect"),
  width  = 14,
  height = 4
)
```

```{r}
rm(EJA_lcwgs_coast_isj_vcf_diff, hi_het_coast_isj, hi_het_coast_isj_plots)
gc()
```

## For North

```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered_north --keep-allele-order --recode vcf --out 8-Embiotoca_filtered_north
```

```{r}
EJA_lcwgs_north_vcf <- read.vcfR(here("plink2", "8-Embiotoca_filtered_north.vcf"))

popmap_lcwgs_north <- read_csv2(here("data", "individuals_coordinates.csv")) |> 
  select(Indiv, Localisation) |>
  rename(id = Indiv, pop = Localisation) |> 
  filter(pop %in% c("BBA", "BIGC", "ELK", "HOP", "SB", "TB", "SCH")) |> 
  data.frame()
```

```{r}
EJA_lcwgs_north_vcf_diff <- vector(mode='list', length=3)
parameters <- c(.5, .75, 1)

for (i in 1:3) {
  EJA_lcwgs_north_vcf_diff[[i]] <- alleleFreqDiff(EJA_lcwgs_north_vcf,
                                                  pm = popmap_lcwgs_north,
                                                  p1 = "BBA", p2 = "SB",
                                                  difference = parameters[i])
}

rm(EJA_lcwgs_north_vcf)
gc()
```

Difference = 1.0 : 2348 sites
Difference = 0.9 : 9364 sites
Difference = 0.8 : 26991 sites (>1% SNPs)
Difference = 0.75 : 38762
Difference = 0.7 : 55736 sites
Difference = 0.5 : 183805 sites

```{r}
hi_het_north <- vector(mode='list', length=3)

for (i in 1:3) {
  hi_het_north[[i]] <- hybridIndex(EJA_lcwgs_north_vcf_diff[[i]],
                                   pm = popmap_lcwgs_north,
                                   p1 = "BBA", p2 = "SB")
}
```

```{r}
#| eval: false

write_rds(hi_het_north, here("output", "Triangle_analysis_north.rds"))
```

```{r}
hi_het_north_plots <- vector(mode='list', length=3)

for (i in 1:3) {
  hi_het_north_plots[[i]] <- hi_het_north[[i]] |>
    mutate(
      pop = factor(
        pop,
        levels = c("BBA", "TB", "SCH", "ELK", "HOP", "BIGC", "SB")
      )
    ) |> 
    arrange(pop) |>
    triangle.plot(colors = turbo(n = 7)) +
    labs(title = paste0("d = ", as.character(parameters[i])))
}

save_open_plot(
  path   = here("output", "plots", "north_Triangle_plots.pdf"),
  plot   = Reduce("+", hi_het_north_plots) +
    plot_layout(axis_titles = "collect", guides = "collect"),
  width  = 14,
  height = 4
)
```


```{r}
rm(EJA_lcwgs_north_vcf_diff, hi_het_north, hi_het_north_plots)
gc()
```

## For South

```{bash}
cd C:/Users/aoules/Desktop/R_projects/Surfperch_GenPop/plink2

./plink2 --bfile 8-Embiotoca_filtered_south --keep-allele-order --recode vcf --out 8-Embiotoca_filtered_south
```

```{r}
EJA_lcwgs_south_vcf <- read.vcfR(here("plink2", "8-Embiotoca_filtered_south.vcf"))

popmap_lcwgs_south <- read_csv2(here("data", "individuals_coordinates.csv")) |> 
  select(Indiv, Localisation) |>
  rename(id = Indiv, pop = Localisation) |> 
  filter(pop %in% c("SB", "PD", "RB", "LAG", "SD")) |> 
  data.frame()
```

```{r}
EJA_lcwgs_south_vcf_diff <- vector(mode='list', length=3)
parameters <- c(.5, .75, 1)

for (i in 1:3) {
  EJA_lcwgs_south_vcf_diff[[i]] <- alleleFreqDiff(EJA_lcwgs_south_vcf,
                                                  pm = popmap_lcwgs_south,
                                                  p1 = "SB", p2 = "SD",
                                                  difference = parameters[i])
}

rm(EJA_lcwgs_south_vcf)
gc()
```

With 1 : 234
.75 : 3367
.5 : 54787

```{r}
hi_het_south <- vector(mode='list', length=3)

for (i in 1:3) {
  hi_het_south[[i]] <- hybridIndex(EJA_lcwgs_south_vcf_diff[[i]],
                                   pm = popmap_lcwgs_south,
                                   p1 = "SB", p2 = "SD")
}
```

```{r}
#| eval: false

write_rds(hi_het_south, here("output", "Triangle_analysis_south.rds"))
```

```{r}
hi_het_south_plots <- vector(mode='list', length=3)

for (i in 1:3) {
  hi_het_south_plots[[i]] <- hi_het_south[[i]] |>
    mutate(
      pop = factor(
        pop,
        levels = c("SB", "PD", "RB", "LAG", "SD")
      )
    ) |> 
    arrange(pop) |>
    triangle.plot(colors = turbo(n = 5)) +
    labs(title = paste0("d = ", as.character(parameters[i])))
}

save_open_plot(
  path   = here("output", "plots", "south_Triangle_plots.pdf"),
  plot   = Reduce("+", hi_het_south_plots) +
    plot_layout(axis_titles = "collect", guides = "collect"),
  width  = 14,
  height = 4
)
```

```{r}
rm(EJA_lcwgs_south_vcf_diff, hi_het_south, hi_het_south_plots)
gc()
```

# Genotype-Environment Association

## Prep

```{r}
#| eval: false

Temperatures <- read_rds(here("data", "Copernicus", "annual_average_SST.rds")) |> 
  terra::unwrap()

Coordinates <- read_csv2(here("data", "individuals_coordinates_temps.csv")) |>
  select(Longitude, Latitude) |> 
  as.data.frame()

Temp_data <- terra::extract(
  x = Temperatures,
  y = data.frame(Coordinates),
  cells = FALSE
)

read_csv2(here("data", "individuals_coordinates_temps.csv")) |> 
  mutate("Temp (Celsius)" = Temp_data$mean) |> 
  write_csv2(here("data", "individuals_coordinates_temps.csv"))

rm(Temperatures, Coordinates, Temp_data)
```

```{r}
ggplot() +
  geom_spatraster(data = Temperatures) +
  geom_point(data = Coordinates, aes(x = Longitude, y = Latitude), shape = 21)
```

## All pops

### Data preparation

```{r}
Temp_data <- read_csv2(here("data", "individuals_coordinates_temps.csv")) |> 
  pull(`Temp (Celsius)`) |> 
  as.matrix()
```


```{r}
Genotype <- LEA::read.geno(here("data", "8-Embiotoca_filtered.geno"))

Genotype_noNA <- Genotype[, !apply(Genotype, 2, function(col) any(col == 9))]

snmf(here("data", "8-Embiotoca_filtered.geno"), K = 7,  project = "new",)

impute(
  object     = load.snmfProject(
    here("data", "snmf", "8-Embiotoca_filtered.snmfProject")
  ),
  input.file = here("data", "8-Embiotoca_filtered.geno"),
  method      = "mode",
  K          = 7,
  run        = 1
)

Genotype_imputed <- read.lfmm(here("data", "8-Embiotoca_filtered.lfmm_imputed.lfmm"))

rm(Genotype)
```

### Analysis

```{r}
mod_lfmm <- LEA::lfmm2(
  input        = Genotype_noNA,
  env          = Temp_data, 
  K            = 4, 
  effect.sizes = TRUE
)
```

```{r}
pv <-  lfmm2.test(mod_lfmm, 
                  input = Genotype_noNA, 
                  env   = Temp_data, 
                  full  = TRUE)

plot(-log10(pv$pvalue), cex = .3, pch = 19, col = "blue")
```

## Coast

### Data preparation

```{r}
#| eval: false

vcf2geno(
  input.file  = here("plink2", "8-Embiotoca_filtered_coastal.vcf"),
  output.file = here("data", "8-Embiotoca_filtered_coastal.geno")
)
```

```{r}
Genotype_coast <- LEA::read.geno(here("data", "8-Embiotoca_filtered_coastal.geno"))

Genotype_coast_noNA <- Genotype_coast[, !apply(Genotype_coast, 2, function(col) any(col == 9))]

rm(Genotype_coast)
```

```{r}
Temp_data_coast <- read_csv2(here("data", "individuals_coordinates_temps.csv")) |> 
  filter(Localisation != "CAT") |>
  filter(Localisation != "GUA") |>
  filter(Localisation != "ISJ") |>
  filter(Localisation != "SCL") |>
  pull(`Temp (Celsius)`) |> 
  as.matrix()
```


### Analysis

```{r}
mod_lfmm_coast <- LEA::lfmm2(
  input        = Genotype_coast_noNA,
  env          = Temp_data_coast, 
  K            = 3,
  effect.sizes = TRUE
)
```

```{r}
pv_coast <- lfmm2.test(
    object = mod_lfmm_coast,
    input  = Genotype_coast_noNA, 
    env    = Temp_data_coast, 
    full   = TRUE
  )

plot(-log10(pv_coast$pvalue), cex = .3, pch = 19, col = "blue")
```